### ISSUE TITLE
Follow-up issues for parsing *_bounds_cast operators.

### ISSUE THREAD

**Issue by dtarditi** [Microsoft/checkedc-clang#260]
It is important to be defensive when programming in a large code base.  People can introduce things like memory corruptions that cause programs to operate incorrectly.  By having defaults and cover all cases (even if we think they are logically impossible), we catch problems quicker and itis easier to debug problems when they do occur.  That means covering all the possible cases for enums in switch statements or conditional tests.  In addition, we should also have `default` cases for switch statements and avoid silently falling off the end of `if`-chains..

- In ASTDumper.cpp around line 1966, I think you should introduce an else statement that handles the situation where Kind does not have an expected value.
```
if (kind == .. Assume)
else if (kind ==  ... Dymaic)
else {
  OS << "Ilegal_bounds_cast_kind"
}
```
It's possible that there would be an error inside the compiler that results in an invalid kind.  We would want to know that when we dump the AST.
- I think a similar check for an unexpected case in StmtPrint.cpp is needed (around line 1907).

I think the `children()` method for the `BoundsCastExpr`class needs to include the base expression (Op, which is returned by CastExpr.getSubExpr()).   Children is supposed to return an iterator over all the child nodes in the AST of the current node.  The definition of the clang IR is somewhat unclear about this.  If you look at all the places where the method is called, it is clear that they expected all AST child nodes to be included

**Comment by dtarditi** [comment]
Also, we should check that the type arguments to bounds_cast operators are pointer types.

**Comment by jijoongmoon** [comment]
When I looked into the VisitCastExpr in ASTDumper, there is also print out about the bounds expression followed by "Inferred" string and if the CastExpr has bounds. and we also have children in CastExpr as Stmt* SubExprs[2], one for the op and one for the bounds. So I wonder that we may do not want to dump for the bounds in ASTDumper.  Unless it prints the bounds twice, one in the ASTDumper, and one in the children iterator.

**Comment by dtarditi** [comment]
Yes, I agree that we should dump the bounds only once for a BoundsCastExpr in ASTDumper.  Given your observation, I would suggest (1) not dumping the inferred bounds for a bound cast expr in VisitCastExpr (2) not doing anything special to dump the bounds for in BoundsCastExpr.   The bounds will be included as one of the children.   Does this seem good?

**Comment by dtarditi** [comment]
I fixed the issue with the children method in BoundsCastExpr with PR #263.

**Comment by dtarditi** [comment]
Addressed by commit 944b36 in PR #269.
